### PR TITLE
Tache4.3_WissemGhouili

### PR DESCRIPTION
**Pull request by WissemGhouili**

### CODE REVIEW

File: scripts/embedding.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    Parameters:
+        X (numpy array or DataFrame): The dataset to transform.
+        n_components (int or None): Number of components to keep. If None, keep all.
+
+    Returns:
+        PCA: Trained PCA model.
+        numpy array: Transformed dataset.
+    """
+    n_components = n_components if n_components else X.shape[1]  # Default: all components
+    pca = PCA(n_components=n_components)
```
Comment (HerySon) [inline scripts/embedding.py:19]: Il faudrait laisser la possibilité à l'utilisateur de pouvoir modifier les paramètres de la PCA, par exemple en laissant ces paramètres en argument optionnels de ta fonction

File: scripts/embedding.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    Determine the optimal number of PCA components to retain.
+
+    Parameters:
+        X (numpy array or DataFrame): The dataset.
+        variance_threshold (float): Minimum cumulative variance to retain.
+
+    Returns:
+        int: Optimal number of components.
+    """
+    pca = PCA().fit(X)
```
Comment (HerySon) [inline scripts/embedding.py:34]: Pas la peine de réentrainer la PCA ici !! Tu l'as déja fait et renvoyée dans ta fonction `apply_pca` : il suffirait de récupérer en entrée de cette fonction ta PCA déja calculée

File: scripts/embedding.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    Parameters:
+        X (numpy array or DataFrame): Feature dataset.
+        y (numpy array or Series): Target variable.
+        k (int): Number of features to keep.
+
+    Returns:
+        numpy array: Reduced dataset with top k features.
+    """
+    selector = SelectKBest(score_func=f_classif, k=k)
+    X_selected = selector.fit_transform(X, y)
```
Comment (HerySon) [inline scripts/embedding.py:70]: Attention, ici il ne faut pas spécifier de y car tu es dans une tâche d'apprentissage non supervisé : tu n'as pas de label

File: scripts/embedding.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    Parameters:
+        X (numpy array or DataFrame): Feature dataset.
+        y (numpy array or Series): Target variable.
+        k (int): Number of features to keep.
+
+    Returns:
+        numpy array: Reduced dataset with top k features.
+    """
+    selector = SelectKBest(score_func=f_classif, k=k)
```
Comment (HerySon) [inline scripts/embedding.py:69]: La méthode de sélection basée sur l'ANOVA est plutôt utilisée pour la classification (voir [la doc](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html#sklearn.feature_selection.f_classif)). Il vaudrait mieux choisir des méthodes non dépendantes du type de tâche

**Review by HerySon — CHANGES_REQUESTED** [review]

**Comment by HerySon** [comment]
@WissemGhouili

File: scripts/embedding.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        pca (PCA): The trained PCA model.
+        X (numpy array or DataFrame): Feature dataset.
+
+    Returns:
+        None
+    """
+    pcs = pca.components_
+    fig, ax = plt.subplots(figsize=(14, 14))
+
+    for i, (x, y) in enumerate(zip(pcs[0, :], pcs[1, :])):
```
Comment (HerySon) [inline scripts/embedding.py:88]: Même si on représente souvent les deux premières composantes, il serait intéressant de laisser le choix à l'utilisateur des composantes à représenter

**Review by HerySon — CHANGES_REQUESTED** [review]

**Comment by HerySon** [comment]
@WissemGhouili
